Update 'Stop a Node' with more draining info

rmloveland opened the pull request:
Fixes https://github.com/cockroachdb/docs/issues/2436
Fixes https://github.com/cockroachdb/docs/issues/2620

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/2671)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/3cecc6d1346ce6598161e6b7edec1780a28d72a3/

rmloveland commented:
@asubiotto direct links to the updated pages for your convenience:

- [Stop a Node](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/3cecc6d1346ce6598161e6b7edec1780a28d72a3/stable/stop-a-node.html)
- [Cluster Settings](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/3cecc6d1346ce6598161e6b7edec1780a28d72a3/dev/cluster-settings.html) (generated from the code for 2.0)

rmloveland commented:
@a-robinson - Alfonso is on holiday for 2 weeks and suggested you as a potential reviewer.  Mind giving these changes a look?

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/898fa26655d461b2abbb0ff25ba783d3baf91030/

a-robinson commented:
Reviewed 3 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[v2.0/cluster-settings.md, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76qFKFDUJ1WoyssIZr:-L76qFKFDUJ1WoyssIZs:b-1uxuss) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/cluster-settings.md#L25)):*
> ```Markdown
> The following settings can be configured without further input from Cockroach Labs:
> 
> {% include settings/v2.0/settings.md %}
> ```

We shouldn't replace these with the entire list -- they were curated to be non-experimental settings that end users could reasonably understand and use on their own without additional documentation or handholding from us. I know it's a pain to keep this up-to-date, but we should either continue trying to do so or change the sentence above to not claim that they can all be configured without our advice.

---

*[v2.0/stop-a-node.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76s2zZhp3XChgpIOjt:-L76s2zZhp3XChgpIOju:b9282h7) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/stop-a-node.md#L19)):*
> ```Markdown
> When you stop a node, it performs the following steps:
> 
> - Finishes in-flight requests. Note that this is a best effort that times out at the `server.shutdown.query_wait` [cluster setting](cluster-settings.html).
> ```

I'd phrase this as "that times out after the duration specified by the `server.shutdown.query_wait` cluster setting"

---

*[v2.0/stop-a-node.md, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76trobaU5xtkPcgQZv:-L76trobaU5xtkPcgQZw:b58780i) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/stop-a-node.md#L21)):*
> ```Markdown
> - Finishes in-flight requests. Note that this is a best effort that times out at the `server.shutdown.query_wait` [cluster setting](cluster-settings.html).
> - Transfers all *range leases* and Raft leadership to other nodes.
> - Gossips its draining state to the cluster, so that other nodes do not try to distribute query planning to the draining node, and no leases are transferred to the draining node.  Note that this is best effort that times out at the `server.shutdown.drain_wait` [cluster setting](cluster-settings.html), so other nodes may not receive the gossip info in time.
> ```

s/best effort/a best effort/

And a similar rephrasing as above would make sense here, too.

---

*[v2.0/stop-a-node.md, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76uJhAEnCOsFf3uxN4:-L76uJhAEnCOsFf3uxN5:bkh2nky) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/stop-a-node.md#L21)):*
> ```Markdown
> - Finishes in-flight requests. Note that this is a best effort that times out at the `server.shutdown.query_wait` [cluster setting](cluster-settings.html).
> - Transfers all *range leases* and Raft leadership to other nodes.
> - Gossips its draining state to the cluster, so that other nodes do not try to distribute query planning to the draining node, and no leases are transferred to the draining node.  Note that this is best effort that times out at the `server.shutdown.drain_wait` [cluster setting](cluster-settings.html), so other nodes may not receive the gossip info in time.
> ```

I'm not positive how effective the "distribute query planning" part actually is (particularly given https://github.com/cockroachdb/cockroach/issues/23601), but that's certainly how we'd like it to work so including it here is probably fine.

Out of curiosity, @solongordon do you know whether other nodes attempt to avoid scheduling distsql processing on draining nodes? Is it purely based on the leaseholder cache?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[v2.0/cluster-settings.md, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76qFKFDUJ1WoyssIZr:-L7AkoXlfMKky6a1yz7L:bmmlybf) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/cluster-settings.md#L25)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

We shouldn't replace these with the entire list -- they were curated to be non-experimental settings that end users could reasonably understand and use on their own without additional documentation or handholding from us. I know it's a pain to keep this up-to-date, but we should either continue trying to do so or change the sentence above to not claim that they can all be configured without our advice.
</blockquote></details>

cc @dt, who did the work to auto-generate this markdown.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[v2.0/stop-a-node.md, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76uJhAEnCOsFf3uxN4:-L7Ap3RUO_HVG7oaNydE:bdcv6z5) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/stop-a-node.md#L21)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'm not positive how effective the "distribute query planning" part actually is (particularly given https://github.com/cockroachdb/cockroach/issues/23601), but that's certainly how we'd like it to work so including it here is probably fine.

Out of curiosity, @solongordon do you know whether other nodes attempt to avoid scheduling distsql processing on draining nodes? Is it purely based on the leaseholder cache?
</blockquote></details>

I haven't seen it in action but it does look like there's logic for avoiding processing on draining nodes: https://github.com/cockroachdb/cockroach/blob/afdf5e12bfdd11054582be7e33280fa185974d9e/pkg/sql/distsql_physical_planner.go#L540-L556

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[v2.0/cluster-settings.md, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76qFKFDUJ1WoyssIZr:-L7Ausy3m4vHgG3trZ_O:b-qq0k8r) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/cluster-settings.md#L25)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

cc @dt, who did the work to auto-generate this markdown.
</blockquote></details>

I think we should put the whole list here and we should remove the sentence above.

Note that above that sentence is a call-out box saying that you should know what you're doing, or talk to us, before changing settings, which I believe adequately sets expectations around these.

For settings that are riskier or whatever, I think we should mention that it in their defined description (we might want to add a longer-form description field).

For extreme cases, we also have the ability to define settings as `Hidden` which excludes them from `SHOW ALL CLUSTER SETTINGS` and this list, though i'm generally in favor of documenting over hiding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-:-L7BEA7KC4i-NTXfKdSE:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jseldess commented on v2.0/stop-a-node.md:
```diff
@@ -14,7 +14,14 @@ For information about permanently removing nodes to downsize a cluster or react
+
+- Finishes in-flight requests. Note that this is a best effort that times out at the `server.shutdown.query_wait` [cluster setting](cluster-settings.html).
+- Transfers all *range leases* and Raft leadership to other nodes.
+- Gossips its draining state to the cluster, so that other nodes do not try to distribute query planning to the draining node, and no leases are transferred to the draining node.  Note that this is best effort that times out at the `server.shutdown.drain_wait` [cluster setting](cluster-settings.html), so other nodes may not receive the gossip info in time.
```
nit: Use only one space after the first period. I know you prefer 2, @rmloveland, but convention in our docs is 1.

jseldess commented on v2.0/cluster-settings.md:
```diff
@@ -20,19 +20,7 @@ They can be updated anytime after a cluster has been started, but only by the `r
-| `sql.trace.log_statement_execute` | On each node, copy all executed statements to the [logging output](debug-and-error-logs.html). | Boolean | `false` |
+{% include settings/v2.0/settings.md %}
 
 <!-- Add this section back in once `system.settings` has been fleshed out.
```
Please remove this old commented-out text.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/94b0f11b30c65bb883518b55e0369797ce7d668d/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/94f88ce6b36ea556370bb7caa90ba9f96dd2b958/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/253660f05806ebc1bd4afb4f3e47c006007163d7/

rmloveland commented:
Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[v2.0/cluster-settings.md, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76qFKFDUJ1WoyssIZr:-L7VkxzbYXn1IKLZEVUa:b-7qlbqe) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/cluster-settings.md#L25)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think we should put the whole list here and we should remove the sentence above.

Note that above that sentence is a call-out box saying that you should know what you're doing, or talk to us, before changing settings, which I believe adequately sets expectations around these.

For settings that are riskier or whatever, I think we should mention that it in their defined description (we might want to add a longer-form description field).

For extreme cases, we also have the ability to define settings as `Hidden` which excludes them from `SHOW ALL CLUSTER SETTINGS` and this list, though i'm generally in favor of documenting over hiding.
</blockquote></details>

Removed the sentence above the list as recommended -  in 898fa266

---

*[v2.0/cluster-settings.md, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76ShomfgGJB8ZJ-AxU-r2-25:-L7Vn6tVn04v7Ceuarfx:b8ickde) ([raw file](https://github.com/cockroachdb/docs/blob/898fa26655d461b2abbb0ff25ba783d3baf91030/v2.0/cluster-settings.md#L25)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Please remove this old commented-out text.
</blockquote></details>

Removed the commented-out text in 94b0f11b

---

*[v2.0/stop-a-node.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76s2zZhp3XChgpIOjt:-L7Vo_JNJEqt8szkbBFs:b6nw5xu) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/stop-a-node.md#L19)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'd phrase this as "that times out after the duration specified by the `server.shutdown.query_wait` cluster setting"
</blockquote></details>

Updated as recommended in 7fc7a2f2

---

*[v2.0/stop-a-node.md, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76trobaU5xtkPcgQZv:-L7VpxteNNEYDDbQAU1H:b-h9zy9o) ([raw file](https://github.com/cockroachdb/docs/blob/3cecc6d1346ce6598161e6b7edec1780a28d72a3/v2.0/stop-a-node.md#L21)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/best effort/a best effort/

And a similar rephrasing as above would make sense here, too.
</blockquote></details>

Thanks - fixed the "a" best effort thing in 94f88ce6

Updated the cluster setting language in 253660f0

---

*[v2.0/stop-a-node.md, line 21 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L76ShomfgGJB8ZJ-AxV-r2-21:-L7VpvefVT20ng5FEO6Z:b5sgqpy) ([raw file](https://github.com/cockroachdb/docs/blob/898fa26655d461b2abbb0ff25ba783d3baf91030/v2.0/stop-a-node.md#L21)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

nit: Use only one space after the first period. I know you prefer 2, @rmloveland, but convention in our docs is 1.
</blockquote></details>

Oops, sorry - fixed in 94f88ce6

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671)*
<!-- Sent from Reviewable.io -->

rmloveland commented:
Hey @a-robinson, at @jseldess' recommendation I'm going to try to add the 1.1.{5,6} changes to this PR as well.  Do you mind answering some additional questions?  I'm a lot less clear on these versions, as you will see from the below. :-)

I have some text below (which I also pushed in 4f5db79a so you can more easily comment) that I prepared in person with Alfonso on FixitDay before he left for holiday. Each bullet is followed by the version in which Alfonso said the changes occurred.

I made it a numbered list here so I can ask related questions for each item below.

First, the draft text from 4f5db79a:

<blockquote>
1. Finishes in-flight requests.  Note that this is a best effort that times out after the duration specified by the `???` cluster setting (1.1.5)
2. Transfers all *range leases* and Raft leadership to other nodes. (1.1.6)
3. Gossips its draining state to the cluster so that no leases are transferred to the draining node. Note that this is a best effort that times out after the duration specified by the `???` cluster setting, so other nodes may not receive the gossip info in time. (1.1.6)
4. No new ranges are transferred to the draining node, to avoid a possible loss of quorum after the node shuts down. (1.1.5)
</blockquote>

My questions:

1. Is there a cluster setting for this in 1.1.6?  I'm not seeing one (at least with the `server.*` prefix) in the cluster settings SQL output below.  If no setting, is there a default we can document?
2. Is this accurate?
3. Same as question 1 re: is there a cluster setting?
4. Is this accurate?

The binary I'm running to look up these cluster settings is v1.1.6:

```
build:      CCL v1.1.6 @ 2018/03/12 17:55:09 (go1.8.3)
```

Cluster settings don't seem to allow for control of the features in items 1 and 3 above, but I'm not very familiar:

```
=> SHOW ALL CLUSTER SETTINGS;

...
| server.consistency_check.interval   | 24h0m0s  | d | the time between range consistency checks; set to 0 to disable consistency checking                    |
| server.declined_reservation_timeout | 1s       | d | the amount of time to consider the store throttled for up-replication after a reservation was declined |
| server.failed_reservation_timeout   | 5s       | d | the amount of time to consider the store throttled for up-replication after a failed reservation call  |
| server.remote_debugging.mode        | local    | s | set to enable remote debugging, localhost-only or disable (any, local, off)                            |
| server.time_until_store_dead        | 5m0s     | d | the time after which if there is no new gossiped information about a store, it is considered dead      |
| server.web_session_timeout          | 168h0m0s | d | the duration that a newly created web session will be valid                                            |
...
```

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/4f5db79ac4fa270c208cb97105b03f9be66ae1fd/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/64e718fa417ecea36c738f0e6850db8585e9973f/

a-robinson commented on v1.1/stop-a-node.md:
```diff
@@ -14,6 +14,11 @@ For information about permanently removing nodes to downsize a cluster or react
 
 ### How It Works
 
+- Finishes in-flight requests.  Note that this is a best effort that times out after the duration specified by the `???` cluster setting (1.1.5)
```
This setting is not in v1.1. The server just cancels all current sessions without waiting.

rmloveland replied:
Thanks, updated to say that in 14139e9b

a-robinson commented on v1.1/stop-a-node.md:
```diff
@@ -14,6 +14,11 @@ For information about permanently removing nodes to downsize a cluster or react
 
+- Finishes in-flight requests.  Note that this is a best effort that times out after the duration specified by the `???` cluster setting (1.1.5)
+- Transfers all *range leases* and Raft leadership to other nodes. (1.1.6)
+- Gossips its draining state to the cluster so that no leases are transferred to the draining node. Note that this is a best effort that times out after the duration specified by the `???` cluster setting, so other nodes may not receive the gossip info in time. (1.1.6)
```
This still happens as of 1.1.6. In 1.1.5 and earlier, this part was broken.

rmloveland replied:
Thanks for confirming.  Do you know if there is a cluster setting for this in 1.1.6?  None of the [documented cluster settings for 1.1.6](https://www.cockroachlabs.com/docs/stable/cluster-settings.html) look to be the one.  And it wasn't clear from a quick `SHOW ALL CLUSTER SETTINGS` on a 1.1.6 binary (though I may have missed it).

a-robinson commented on v1.1/stop-a-node.md:
```diff
@@ -14,6 +14,11 @@ For information about permanently removing nodes to downsize a cluster or react
 ### How It Works
 
+- Finishes in-flight requests.  Note that this is a best effort that times out after the duration specified by the `???` cluster setting (1.1.5)
+- Transfers all *range leases* and Raft leadership to other nodes. (1.1.6)
```
This still happens as of 1.1.6. In 1.1.5 and earlier, these parts didn't always happen as intended.

rmloveland replied:
Great, thanks!  Since the stable web docs show 1.1.6 now, I'll leave this in (minus the version number).

a-robinson commented on v1.1/stop-a-node.md:
```diff
@@ -14,6 +14,11 @@ For information about permanently removing nodes to downsize a cluster or react
+- Finishes in-flight requests.  Note that this is a best effort that times out after the duration specified by the `???` cluster setting (1.1.5)
+- Transfers all *range leases* and Raft leadership to other nodes. (1.1.6)
+- Gossips its draining state to the cluster so that no leases are transferred to the draining node. Note that this is a best effort that times out after the duration specified by the `???` cluster setting, so other nodes may not receive the gossip info in time. (1.1.6)
+- No new ranges are transferred to the draining node, to avoid a possible loss of quorum after the node shuts down. (1.1.5)
```
This is true for all 1.1.x versions, as far as I'm aware.

rmloveland replied:
Thanks! Again, leaving in but removing the version number.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/14139e9bf4e8204750d450ab73f9c1fcc31d5a48/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/76a56222df363f3bf2bf63b0ee883342dc258b11/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/6c72d46e92a5659fe6d70e18e2bbc618b573fc25/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/d9d22a13efe08cf98fccad6ef9c44b846cf9c9d9/

a-robinson commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r5, 1 of 1 files at r8, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[v1.1/stop-a-node.md, line 19 at r9](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-L7_53JuealOANKWzMv5-r9-19:-L7fFOokbhO-wRRk58V2:b7rvqkc) ([raw file](https://github.com/cockroachdb/docs/blob/64e718fa417ecea36c738f0e6850db8585e9973f/v1.1/stop-a-node.md#L19)):*
<details><summary><i>Previously, rmloveland (Rich Loveland) wrote…</i></summary><blockquote>

Thanks for confirming.  Do you know if there is a cluster setting for this in 1.1.6?  None of the [documented cluster settings for 1.1.6](https://www.cockroachlabs.com/docs/stable/cluster-settings.html) look to be the one.  And it wasn't clear from a quick `SHOW ALL CLUSTER SETTINGS` on a 1.1.6 binary (though I may have missed it).
</blockquote></details>

Nope, no cluster setting in any of the v1.1 releases. It'd have to be something pretty critical for us to introduce/backport a new cluster setting in a patch release.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2671#-:-L7fHXtrjSsRIwYVR0P0:b-uwkowd)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/094d164c56494a2366787b15985231649a0e6861/

rmloveland commented:
@jseldess do you want a final editorial look at the v1.1 changes before merge?

jseldess commented on v1.1/stop-a-node.md:
```diff
@@ -14,7 +14,12 @@ For information about permanently removing nodes to downsize a cluster or react
 
-When you stop a node, CockroachDB lets the node finish in-flight requests and transfers all **range leases** off the node before shutting it down. If the node then stays offline for a certain amount of time (5 minutes by default), the cluster considers the node dead and starts to transfer its **range replicas** to other nodes as well.
+- Cancels all current sessions without waiting.
+- Transfers all *range leases* and Raft leadership to other nodes.
```
nit: bold instead of italics.

rmloveland replied:
Thanks, fixed in bb14e1bc

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/bb14e1bcae2a9bdd83f806436b1d732ef75baf25/
